File: app/src/main/java/com/bubelov/coins/ui/activity/ExchangeRatesActivity.java
package com.bubelov.coins.ui.activity;

import android.os.AsyncTask;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.widget.ViewSwitcher;

import com.bubelov.coins.R;
import com.bubelov.coins.data.repository.rate.ExchangeRatesRepository;
import com.bubelov.coins.domain.ExchangeRate;
import com.bubelov.coins.ui.adapter.ExchangeRatesAdapter;

import java.util.Collection;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * @author Igor Bubelov
 */

public class ExchangeRatesActivity extends AbstractActivity {
    @BindView(R.id.toolbar)
    Toolbar toolbar;

    @BindView(R.id.state_switcher)
    ViewSwitcher stateSwitcher;

    @BindView(R.id.list)
    RecyclerView ratesView;

    @Inject
    ExchangeRatesRepository exchangeRatesRepository;

    private ExchangeRatesAdapter ratesAdapter;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        dependencies().inject(this);
        setContentView(R.layout.activity_exchange_rates);
        ButterKnife.bind(this);

        toolbar.setNavigationOnClickListener(v -> supportFinishAfterTransition());

        ratesView.setLayoutManager(new LinearLayoutManager(this));
        ratesView.setHasFixedSize(true);

        ratesAdapter = new ExchangeRatesAdapter();
        ratesView.setAdapter(ratesAdapter);

        new LoadExchangeRatesTask().execute();
    }

    private void setLoading(boolean loading) {
        stateSwitcher.setDisplayedChild(loading ? 1 : 0);
    }

    private class LoadExchangeRatesTask extends AsyncTask<Void, Void, Collection<ExchangeRate>> {
        @Override
        protected void onPreExecute() {
            setLoading(true);
            ratesAdapter.getItems().clear();
            ratesAdapter.notifyDataSetChanged();
        }

        @Override
        protected Collection<ExchangeRate> doInBackground(Void... voids) {
            return exchangeRatesRepository.getExchangeRates();
        }

        @Override
        protected void onPostExecute(Collection<ExchangeRate> exchangeRates) {
            ratesAdapter.getItems().addAll(exchangeRates);
            ratesAdapter.notifyDataSetChanged();
            setLoading(false);
        }
    }
}

File: app/src/main/java/com/bubelov/coins/ui/activity/ExchangeRatesActivity.kt
/*
 * This is free and unencumbered software released into the public domain.
 *
 * Anyone is free to copy, modify, publish, use, compile, sell, or
 * distribute this software, either in source code form or as a compiled
 * binary, for any purpose, commercial or non-commercial, and by any
 * means.
 *
 * In jurisdictions that recognize copyright laws, the author or authors
 * of this software dedicate any and all copyright interest in the
 * software to the public domain. We make this dedication for the benefit
 * of the public at large and to the detriment of our heirs and
 * successors. We intend this dedication to be an overt act of
 * relinquishment in perpetuity of all present and future rights to this
 * software under copyright law.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
 * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 *
 * For more information, please refer to <https://unlicense.org>
 */

package com.bubelov.coins.ui.activity

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager

import com.bubelov.coins.R
import com.bubelov.coins.model.CurrencyPair
import com.bubelov.coins.ui.adapter.ExchangeRatesAdapter

import com.bubelov.coins.ui.viewmodel.ExchangeRatesViewModel
import kotlinx.android.synthetic.main.activity_exchange_rates.*
import org.jetbrains.anko.selector

class ExchangeRatesActivity : AppCompatActivity() {
    private lateinit var model: ExchangeRatesViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_exchange_rates)
        model = ViewModelProviders.of(this).get(ExchangeRatesViewModel::class.java)

        toolbar.apply {
            setNavigationOnClickListener { supportFinishAfterTransition() }
            inflateMenu(R.menu.exchange_rates)

            setOnMenuItemClickListener {
                if (it.itemId == R.id.currency) {
                    selector(title = getString(R.string.currency), items = CurrencyPair.values().map { it.displayCurrency }, onClick = { _, index ->
                        val pair = CurrencyPair.values()[index]
                        model.pair.value = pair
                        model.analytics.logEvent("change_exchange_rates_currency_pair", pair.toString())
                    })
                }

                true
            }
        }

        ratesView.layoutManager = LinearLayoutManager(this)
        ratesView.setHasFixedSize(true)
        val ratesAdapter = ExchangeRatesAdapter()
        ratesView.adapter = ratesAdapter

        model.pair.observe(this, Observer {
            if (it != null) {
                toolbar.menu.findItem(R.id.currency).title = it.displayCurrency
            }
        })

        model.rates.observe(this, Observer { rates ->
            if (rates != null) {
                ratesAdapter.items = rates
                ratesAdapter.notifyDataSetChanged()
            }
        })

        model.pair.value = CurrencyPair.BTC_USD
    }
}